crates/typst-utils/src/format.rs

```diff
@@ -0,0 +1,136 @@
+//! Different formatting functions.
+
+use std::time::Duration;
+
+/// Returns value with `n` digits after floating point where `n` is `precision`.
+/// Standard rounding rule applies (if `n+1`th digit >= 5 then round up).
+/// If `value` is +/- infinity returns `value`.
+///
+/// Will panic for a very large `value` and `precision` (if the `f64` limit is hit).
+/// In practice, it should never panic, otherwise the implementation have to be
+/// more robust.
+pub fn round_with_precision(value: f64, precision: u8) -> f64 {
+    let offset = 10_f64.powi(precision.into());
+    if [f64::NEG_INFINITY, f64::INFINITY].contains(&value) {
+        return value;
+    }
+    assert!(
+        value * offset < f64::MAX,
+        "Hit the upper limit of f64 when rounding! ({value} * 10^{precision})"
+    );
```

Comment:
I think panicking here is problematic since this can be called from Typst formatting code.

Reply:
What is "Typst formatting code"?

Reply:
I meant `format_float` which is called by `repr`, so this could be hit by user code.

Reply:
Yeah, I figured that one out already. https://github.com/typst/typst/pull/4587#discussion_r1718507312
Since the `.typ` tests failed. The point is that this assertion in practice never should be triggered. But if you want, I can try handling large numbers somehow, not yet sure how.

Reply:
I think we can explicitly mention in the function's docstring that accuracy isn't guaranteed for large enough numbers. In practice, this function is only used to display debugging or user-friendly information, so it doesn't matter too much if there's some imprecision here. A different function would have to be used for actual calculations (instead of just for string formatting purposes), as I see it.

Reply:
If the value is already huge, there is probably not float anyway that can represent it with the desired rounding. So can't we just return it as-is?

Reply:
You can have numbers with 308 zeros. So I think that you definitely can have a huge number and still get precise rounding to some degree. Leaving the algorithm as is means that in some cases numbers will convert to infinity. Not great. Leaving huge numbers as is... Sounds like the function just gives up on it's job.

I just got an idea. Why not just separate integer and fractional parts and then perform the rounding. Then add the two and you're done. Last operation can affect precision, but that will be on the user, IMO. So as a result, no assertions required.

Reply:
That might work as well. I just mean that I would always prefer the rounding to not happen rather than the assertion to trigger.

Reply:
Apparently in Rust `123.455 - 123 = 0.4549999999999983` which failed the test: https://github.com/Andrew15-5/typst/actions/runs/10420114994/job/28859494612. Tests already do their job, which is nice.

Reply:
I asked rustaceans and the fix is to use other data types. Fortunately, adding [rust_decimal](https://crates.io/crates/rust_decimal) dependency not only solves the problem, but also adds only 1 crate. I see this as an absolute win!

Reply:
I don't think the dependency pulls its weight here; as mentioned before, precision isn't important here. I'd rather follow Laurenz's approach of just returning the float as is if it's too large.

Do we know in which cases we can cause any sort of "overflow" in the first place? Would be nice to have some tests with really large floats first and see how the function behaves, and adjust from there.

Reply:
I'll think about it later.

Reply:
> Do we know in which cases we can cause any sort of "overflow" in the first place?

When `value * 10_f64.powi(precision.into())` `> f64::MAX` or `< f64::MIN`.

Reply:
In the answer [here](https://stackoverflow.com/a/28656825) and from the mantissa, that is 52(+1) bits, the largest integer is 9e15. Going above will cause rounding and conversion error. So, if an input value is a concrete number without any errors, then numbers above 2^53 are none of our concern. Since you can't store fractional part for such big numbers, the "rounding with precision" is useless and therefore can be skipped by returning the input value.

This also means that for fractional, part you can only precisely represent up to `log10(2^53)` digits.

Now, even if you round integers to 2^60 and precision to 20, multiplication will still give you a finite number. If I'm not missing anything, this means that you don't need to explain any of this in the docs as it will be just a remix of IEEE 754 standard. Problem solved.